Various performance and memory usage optimizations, cutting down initialization time in ~half

fabiospampinato opened the pull request:
With the optimizations contained in this PR the time required to parse `dictionary-fr` went down from this:

![image](https://user-images.githubusercontent.com/1812093/93771500-7ffcad00-fc15-11ea-8e30-9ad2edf3bb34.png)

To this:

![image](https://user-images.githubusercontent.com/1812093/93771509-82f79d80-fc15-11ea-9dec-c14aea3a8b7b.png)

This closes #31, as serializing parsed dictionaries doesn't really seem to be a viable strategy, but addressing some of the performance issues lead to significant improvements. There probably is still some room for improvement but I ran out of ideas, building an object with millions of properties just seems inherently pretty expensive.

wooorm commented on lib/util/add.js:
```diff
@@ -6,6 +6,24 @@ module.exports = add
+  if (word in dict) {
+    var curr = dict[word]
+    if (curr === NO_RULES) {
+      dict[word] = rules.slice(0)
```
is `.slice(0)` really faster than `concat()`? I kinda like the latter style and I’m not sure why the former would be faster?

fabiospampinato replied:
I think it was marginally faster, but probably the range of variation in the benchmark is greater than so I may have misinterpreted the result, I can't remember now what difference that made, I can try both again 👍

wooorm replied:
I’m assuming V8 can deal with both as fast, or maybe a marginal difference indeed. In which case I’d prefer concat

fabiospampinato replied:
The first 3 runs are with `concat`, the next 3 are with `slice(0)`, the remaining 3 are with `concat` 🤷‍♂️ 

![image](https://user-images.githubusercontent.com/1812093/93782704-c7d60100-fc22-11ea-9c52-609624c37cdc.png)

Let's go with concat.

wooorm commented on lib/util/affix.js:
```diff
@@ -235,7 +235,7 @@ function affix(aff) {
       flags[ruleType] = value
     } else if (ruleType === keyType) {
-      flags[ruleType] = flags[ruleType].concat(parts[1].split('|'))
+      flags[ruleType].push.apply(flags[ruleType], parts[1].split('|'))
```
Can `push.apply` be moved to the top, similar to `own = ...`?

As in:

```
var push = [].push
```

And then:

```suggestion
      push.apply(flags[ruleType], parts[1].split('|'))
```

fabiospampinato replied:
Sure. Actually I want to try optimizing that further, maybe always growing the larger array of the two would be beneficial, but I don't think nspell usually generates long-enough arrays for that to make a difference.

wooorm replied:
Flags typically aren’t very long, but there are thousands of entries which include 1-5 of ’em.

wooorm commented on lib/util/rule-codes.js:
```diff
@@ -2,23 +2,25 @@
     index = 0
     length = value.length
 
+    result = new Array(Math.ceil(length / 2))
```
```suggestion
    result = []
```

Is the current needed? Or can it be simpler?

fabiospampinato replied:
Those two lines are logically equivalent in the context of how that function works, i.e. they will always lead to the same result, but using `new Array` preallocates an array of the right size immediately, while growing a 0-length one will be more expensive, both because its size will need to be changed at runtime and because v8 may overshoot it and think that it will actually grow to contain 10 items (maybe because it went from 0-length to 6-length quickly) while in reality it will stop at 6 or something.

wooorm replied:
If it is a significant improvement, can you add a comment above that line stating that it shaved off x ms or so?

fabiospampinato replied:
According to my understanding of how stuff like v8 work that's an improvement for sure, I can add a comment about that but attaching a precise number of ms that that shaved off would be challenging, there are too many factors that go into each benchmark execution.

fabiospampinato replied:
First 3 runs with `[]`, remaining 3 runs with `new Array`:

![image](https://user-images.githubusercontent.com/1812093/93786618-6ebc9c00-fc27-11ea-804d-3eccc14e9699.png)

I'm not sure how much one should see into this though, I should be running benchmarks like a million times to get something statistically significant proabably.

wooorm reviewed (commented):
Awesome work! Couple Qs!

fabiospampinato commented:
For future reference here's the command for benchmarking `dictionary-fr`:

```
node -e "require('dictionary-fr')((err,dict)=>{const nspell=require('.');console.time('bench');nspell(dict);console.timeEnd('bench');})"
```

fabiospampinato commented:
@wooorm I've addressed all the comments and squeezed in a few more minor optimizations. I've tried to make it significantly faster still but that doesn't seems impossible, something like 80% of the work in being spent on this line for some reason:

```
if (!(newWord in dict)) dict[newWord] = NO_RULES
```

I'm not sure there's anything that we can do at the JS level to make that faster 🤷 I've filed a [ticket](https://bugs.chromium.org/p/v8/issues/detail?id=10944) with v8 though because I think it is taking too long to construct these objects, Safari seems to build some similar plain objects about 2x fast.

wooorm reviewed (commented):
code looks good

wooorm commented:
npm publish seems to be failing for me. I’ll try again later.

fabiospampinato commented:
I might have broken it:

![image](https://user-images.githubusercontent.com/1812093/93816352-4ea0d300-fc4f-11ea-8322-4a23e6eb3fa1.png)

PR coming 👍
